nibabel/filebasedimages.py

```diff
@@ -511,3 +512,79 @@ def path_maybe_image(klass, filename, sniff=None, sniff_max=1024):
         if sniff is None or len(sniff[0]) < klass._meta_sniff_len:
             return False, sniff
         return klass.header_class.may_contain_header(sniff[0]), sniff
+
+
+class SerializableImage(FileBasedImage):
+    '''
+    Abstract image class for (de)serializing images to/from byte strings.
+
+    The class doesn't define any image properties.
+
+    It has:
+
+    methods:
+
+       * .to_bytes() - serialize image to byte string
+
+    classmethods:
+
+       * from_bytes(bytestring) - make instance by deserializing a byte string
+
+    The following properties should hold:
+
+      * ``klass.from_bytes(bstr).to_bytes() == bstr``
+      * if ``img = orig.__class__.from_bytes(orig.to_bytes())``, then
+        ``img.header == orig.header`` and ``img.get_data() == orig.get_data()``
+
+    Further, for images that are single files on disk, the following methods of loading
+    the image must be equivalent:
+
+        img = klass.from_filename(fname)
+
+        with open(fname, 'rb') as fobj:
+            img = klass.from_bytes(fobj.read())
+
+    And the following methods of saving a file must be equivalent:
+
+        img.to_filename(fname)
+
+        with open(fname, 'wb') as fobj:
+            fobj.write(img.to_bytes())
+
+    Images that consist of separate header and data files will generally
+    place the header with the data, but if the header is not of fixed
+    size and does not define its own size, a new format may need to be
+    defined.
+    '''
+    @classmethod
+    def from_bytes(klass, bytestring):
+        """ Construct image from a byte string
+
+        Class method
+
+        Parameters
+        ----------
+        bstring : bytes
+            Byte string containing the on-disk representation of an image
+        """
+        bio = io.BytesIO(bytestring)
+        file_map = klass.make_file_map({'image': bio, 'header': bio})
```

Comment:
Ouch - this is a bit frightening.   I guess we now have to worry about the order in which the header and image get written, within the image classes?

Reply:
Yes. At present this works for single-file images, where they are already passed a single file object in both fields. I would expect that, if we extend to multi-file images, it would be the responsibility of the implementor to ensure that the serialization is consistent and can be de-serialized.

I'm a little hesitant to pre-engineer for those cases, as I don't have the time or need to work out those problems. As long as we're not providing this function for multi-image files, I'm not sure it's a problem.

Reply:
To be clear, if it looks like the decisions I've made now are going to actively make it harder to support those cases in the future than some other implementation, I'm happy to change it.